modules/internal/ChapelLocale.chpl

```diff
@@ -90,20 +90,142 @@ module ChapelLocale {
     return (subloc != c_sublocid_none
             && subloc != c_sublocid_any
             && subloc != c_sublocid_all);
+  
+  /*
+    regular: Has a concrete BaseLocale instance
+    nilLocale: The _instance is set to nil. Used during setup. Also, as a
+               sentinel value in locale tree operations
+    dummy: Used during initialization for `here` before it is properly setup
+    default: Used to store the default locale instance. Initially set to nil,
+             then "fixed" by LocalesArray to Locales[0]
+    any: Placeholder to represent the notion of "anywhere"
+   */
+  pragma "no doc"
+  enum localeKind { regular, any, nilLocale, dummy, default };
+
+  const nilLocale = new locale(localeKind.nilLocale);
+  var defaultLocale = new locale(localeKind.default);
```

Comment:
It took me awhile to realize that this was probably calling `new _locale` through compiler rewriting.  If we were to rename the record from `_locale` to `locale` and update the strcmp against `_locale` in the compiler to `locale`, would things still compile or break?  (I'm guessing they'd break due to a syntax error in the definition of the record?).  If these lines were rewritten to `new _locale` would they still work?  (that said, I realize there are a *lot* of such lines in this PR).

Reply:
Among your comments, this is the one that I am most hesitant about exploring at this point. Although I agree that we probably need a better strategy for those things. Because we use this same method for many things in the compiler: https://github.com/chapel-lang/chapel/blob/master/compiler/AST/build.cpp#L1563-L1583. OTOH, I can't tell whether you see `locale` as less of a primitive than, say, `string`. And that whether you're fine with `record _string` but not so much with `record _locale`.

One option that I can think of for the near future is to see if we can change the parser to recognize `record locale` in a way that it doesn't emit `TLOCALE` for the second word there and see if we can special-case these variables that way. Another way *might* be is to use pragmas like `locale record` to signal the compiler to behave differently for the `record locale` definition. But I am not sure if that's possible, or how much effort it is to make sure that we don't do anything special for the word `locale` there until we understand the pragma above and do something about it during compilation.

Reply:
It's fine with me if we put it off, and the analogy to string is an apt one (I don't have a bone to pick with strings, but also haven't looked at their code much; I _have_ long since wanted to rename `_domain` and `_array` to something more like `domain` and `array`).

I mostly went through a lot of churn thinking "`new locale()` OK, what's that calling?" (searches for `record locale`, `class locale` finding nothing), "...huh..."  This specific pair were probably particularly bad since they came right near the definition of `record _locale`.  For the time being, how about just adding a comment for these cases saying "`new locale()` becomes `new _locale()` through compiler magic" (or if you have a better explanation than "magic", that's even better.  :)  And/or a comment on `record _locale` that says "The compiler will effectively treat this as `record locale` so that grepping for "record locale" will hit the comment?

I think if we were to special-case the parser in the future, I think we'd probably want to add a `TRECORD TLOCALE` rule that was a compiler error if it occurred anywhere other than the well-defined `ChapelLocale.chpl` module.  I'm also open to a "rename me to locale" pragma or something like that, though I like the idea of the parser change better in terms of minimizing renamings and white lies.